Reworked caching for proxies

Toflar opened the pull request:
As discussed in https://github.com/contao/contao/issues/370#issuecomment-470593637

Toflar committed:
Reworked caching for proxies

Toflar commented:
Ping @ausi @aschempp

ausi commented on core-bundle/src/Resources/contao/classes/FrontendTemplate.php:
```diff
@@ -348,29 +348,38 @@ private function setCacheHeaders(Response $response)
-		// TODO: Add support for proxies so they can vary on member context
-		if (FE_USER_LOGGED_IN === true || BE_USER_LOGGED_IN === true || $objPage->protected || $this->hasAuthenticatedBackendUser())
+		// Do not cache response if front end user is logged in UNLESS explicitly set in the page settings
+		if (FE_USER_LOGGED_IN === true && !$objPage->forceCache)
```
IMO this should just be

```suggestion
		if (0 !== \count($response->headers->getCookies()))
```

So that we don’t store anything in the cache if it was triggered by a cookied request.

ausi commented:
How about something like that:

```php
// Do not cache response if caching was not configured at all
if (($objPage->cache === false || $objPage->cache < 1) && ($objPage->clientCache === false || $objPage->clientCache < 1))
{
	return $this->makeResponsePrivate($response);
}

// Do not cache response if any cookies were received or sent
if (0 !== \count($response->headers->getCookies()) || 0 !== \count(System::getContainer()->get('request_stack')->getCurrentRequest()->headers->getCookies()))
{
	return $this->makeResponsePrivate($response);
}

// Now we have cacheable requests. We do vary on Cookie for those responses that have **NO** cookies set which will
// cause reverse proxies to never load a response from the cache if the request contains a cookie.
// We only do this if you did not configure the page to **EXPLICITLY** cache even though a cookie is present.
if (!$objPage->forceCache) {
	$response->setVary(['Cookie']);
}
```

Toflar commented:
@ausi please review again, I found other issues and reworked the comments. The varying is now also only applied in the shared cache section because it's irrelevant for private caches.

Toflar committed:
More comments. Comments are super useful, especially for this type of complex stuff.

ausi commented on core-bundle/src/Resources/contao/classes/FrontendTemplate.php:
```diff
@@ -356,24 +357,48 @@ private function setCacheHeaders(Response $response)
-		// TODO: Add support for proxies so they can vary on member context
-		if (FE_USER_LOGGED_IN === true || BE_USER_LOGGED_IN === true || $objPage->protected || $this->hasAuthenticatedBackendUser())
+		// Do not cache response if any cookies were received or sent and make sure the response is private.
+		if (0 !== \count($response->headers->getCookies()) || 0 !== \count(System::getContainer()->get('request_stack')->getCurrentRequest()->headers->getCookies()))
```
```suggestion
		if (0 !== \count($response->headers->getCookies()) || 0 !== System::getContainer()->get('request_stack')->getCurrentRequest()->cookies->count())
```

ausi replied:
Is `getCurrentRequest()` the right one here?

Toflar replied:
Imho yes.

aschempp commented on core-bundle/src/Resources/contao/classes/FrontendTemplate.php:
```diff
@@ -348,6 +348,7 @@ private function setCacheHeaders(Response $response)
+		// Do not cache response if caching was not configured at all and make sure the response is private.
 		if (($objPage->cache === false || $objPage->cache < 1) && ($objPage->clientCache === false || $objPage->clientCache < 1))
 		{
 			$response->headers->addCacheControlDirective('no-cache');
```
I think we should replace this with `$response->headers->set('Cache-Control', 'no-cache, no-store');`

Toflar replied:
Done

aschempp commented on core-bundle/src/Resources/contao/classes/FrontendTemplate.php:
```diff
@@ -356,24 +357,48 @@ private function setCacheHeaders(Response $response)
+		if (0 !== \count($response->headers->getCookies()) || 0 !== \count(System::getContainer()->get('request_stack')->getCurrentRequest()->headers->getCookies()))
 		{
 			$response->headers->addCacheControlDirective('no-cache');
 			$response->headers->addCacheControlDirective('no-store');
```
same as above

ausi commented on core-bundle/src/Resources/contao/languages/en/tl_page.xlf:
```diff
@@ -176,6 +176,12 @@
+        <source>Force server cache for logged-in members</source>
+      </trans-unit>
+      <trans-unit id="tl_page.forceCache.1">
+        <source>By default, the server cache is disabled for logged-in members. You can force caching with this checkbox (ATTENTION: This can lead to information leak between logged in members, only do this if you know what you're doing!).</source>
```
> ATTENTION: This can lead to information leak between logged in members

IMO this is not correct. We only store responses in the cache if no cookies are involved regardless of the `forceCache` setting.

The only thing that can happen if `forceCache` gets used incorrectly is that a logged in member gets the public response from the cache instead of a private one personalized to him.

Toflar replied:
True, fixed in e4c289aac71ca8875703b01891eb33bc4eb22376.

aschempp commented on core-bundle/src/Resources/contao/models/PageModel.php:
```diff
@@ -53,6 +53,7 @@
  * @property integer $layout
  * @property boolean $includeCache
  * @property integer $cache
+ * @property integer $forceCache
```
I guess this should be boolean?

Toflar replied:
True fixed in 430dda656dcef832125c1ae6c0b7fa6a231e641d.

aschempp commented on core-bundle/src/Resources/contao/models/PageModel.php:
```diff
@@ -858,6 +859,7 @@ public function loadDetails()
 		$this->groups = $this->protected ? StringUtil::deserialize($this->groups) : false;
 		$this->layout = $this->includeLayout ? $this->layout : false;
 		$this->cache = $this->includeCache ? $this->cache : false;
+		$this->forceCache = $this->forceCache ? $this->forceCache : false;
```
`(bool) $this->forceCache`

Toflar replied:
True fixed in 430dda656dcef832125c1ae6c0b7fa6a231e641d.

aschempp commented on core-bundle/src/Resources/contao/classes/FrontendTemplate.php:
```diff
@@ -356,24 +357,48 @@ private function setCacheHeaders(Response $response)
-		// TODO: Add support for proxies so they can vary on member context
-		if (FE_USER_LOGGED_IN === true || BE_USER_LOGGED_IN === true || $objPage->protected || $this->hasAuthenticatedBackendUser())
+		// Do not cache response if any cookies were received or sent and make sure the response is private.
+		if (0 !== \count($response->headers->getCookies()) || 0 !== \count(System::getContainer()->get('request_stack')->getCurrentRequest()->headers->getCookies()))
```
as usualy, very unsure if `getCurrentRequest` is correct, as the cookies are only on the master request and (probably) not added to subrequests.

Toflar replied:
As usually, it is correct :) The master request is very, very rarely the right one. Cookies can also be set on a subrequest and we have to have the very same handling there. Imagine you create an ESI fragment that uses `Template->getResponse()`. It should work just the same.

aschempp replied:
I'm not sure I understand this correctly.
 1. if Symfony's `HttpCache` is used, the *current* request might be a subrequest for ESI, which means cookies are likely not there (because its not a clone of the master request). So using the master request would be better.
 2. if this is a Varnish or other reverse proxy, it's always the master request, so using the master request would be fine?

aschempp reviewed (commented):
I also wonder if we should not check for cookies, but rather check for the security user and/or if a session was started? This way we would only check for session cookies and ignore all others.

Toflar committed:
Improved description of setting

ausi committed:
Use built-in count method

Co-Authored-By: Toflar <yanick.witschi@terminal42.ch>

Toflar committed:
Minor fixes

Toflar commented:
> I also wonder if we should not check for cookies, but rather check for the security user and/or if a session was started? This way we would only check for session cookies and ignore all others.

Which would be wrong. That's why don't do the session cookie. You don't know what the developers use the cookies for. They might use them to store filter settings incorrectly in which case we also have to make sure the response is not stored :)

Toflar committed:
Removed obsolete method override

Toflar commented:
All comments addressed. Looks good to me now :)

ausi commented on core-bundle/src/Resources/contao/models/PageModel.php:
```diff
@@ -858,6 +859,7 @@ public function loadDetails()
 		$this->groups = $this->protected ? StringUtil::deserialize($this->groups) : false;
 		$this->layout = $this->includeLayout ? $this->layout : false;
 		$this->cache = $this->includeCache ? $this->cache : false;
+		$this->forceCache = (bool) $this->forceCache ? $this->forceCache : false;
```
```suggestion
		$this->forceCache = (bool) $this->forceCache;
```

or:

```suggestion
		$this->forceCache = (bool) ($this->forceCache ? $this->forceCache : false);
```

ausi committed:
Fixed boolean casting

Co-Authored-By: Toflar <yanick.witschi@terminal42.ch>

ausi commented:
Should we also check for authorization headers like `$request->getUserInfo()` or similar?

Toflar commented:
In fact, personally, I wouldn't check any request data at all. It's the task of the reverse proxy not to cache responses to private requests. E.g. Symfony's HttpCache does that by default: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php#L476

So our code is useless for the `HttpCache` case. But it's probably not for Varnish and Co.

ausi commented:
```php
if ($this->isPrivateRequest($request) && !$response->headers->hasCacheControlDirective('public')) {
```

If I read that correctly, private requests do not get cached only if `Cache-Control` is not set to `public`. But we do set it to public in Contao which means we should probably also check for authorization headers in addition to cookies?

Toflar commented:
Yeah we might want to do that. The `HttpCache` does strange things...

ausi reviewed (approved):
Apart from the missing check for authorization headers this looks good to me. 👍

Toflar committed:
Also consider authorization header

Toflar commented:
Added now

leofeyer commented on core-bundle/src/Resources/contao/classes/FrontendTemplate.php:
```diff
@@ -348,32 +348,55 @@ private function setCacheHeaders(Response $response)
 		{
-			$response->headers->addCacheControlDirective('no-cache');
-			$response->headers->addCacheControlDirective('no-store');
+			$response->headers->set('Cache-Control', 'no-cache, no-store');
```
Is this any different from `addCacheControlDirective()`?

Toflar replied:
It's a) shorter and b) it overrides any existing definition which is good.

leofeyer commented on core-bundle/src/Resources/contao/classes/FrontendTemplate.php:
```diff
@@ -386,6 +409,29 @@ private function setCacheHeaders(Response $response)
 		return $response;
 	}
+
+	private function isCurrentRequestPrivate()
```
`: bool`

leofeyer commented:
I think "Force server cache for logged-in members" is not very clear to non-technical users. I think the title and description should be something like "Always load from cache" and "Always load this page from cache, even if a member is logged in. (Note that you can no longer personalize the page for logged in members in this case.)".

Should we also rename the database field to `alwaysLoadFromCache`?

Toflar committed:
Various naming improvements

Toflar commented:
Fine with me, I've addressed your comments. I've also adjusted the labels to `Private cache` and `Shared cache` so they are techincally correct. I left `(client cache)` and `(server cache)` in brackets as additional hint for those who don't know the technically correct terms.

aschempp commented on core-bundle/src/Resources/contao/classes/FrontendTemplate.php:
```diff
@@ -348,32 +348,55 @@ private function setCacheHeaders(Response $response)
-			$response->headers->addCacheControlDirective('no-store');
+			$response->headers->set('Cache-Control', 'no-cache, no-store');
 
 			return $response->setPrivate();
```
I think the `$objPage->clientCache` should also be applied here. If I configure a page in the tree to be user-cacheable, it can still be customized, right?

aschempp commented:
> > I also wonder if we should not check for cookies, but rather check for the security user and/or if a session was started? This way we would only check for session cookies and ignore all others.
> 
> Which would be wrong. That's why don't do the session cookie. You don't know what the developers use the cookies for. They might use them to store filter settings incorrectly in which case we also have to make sure the response is not stored :)

I agree we don't know what the devs do. But they **should** not use a custom cookie for personalization, that's what the session is for, isn't that the case? That's why Symfony makes responses private if they have a session, and not if they have any cookie?

Toflar commented:
> That's why Symfony makes responses private if they have a session, and not if they have any cookie?

That's because they know this is the only thing in Symfony that might be problematic. We don't know so I would stick with the current solution.

aschempp commented:
I remember we had this discussion when we overrided the Symfony listener in Contao 4.4 etc. Maybe we need to discuss the topic again 😞

Toflar commented:
@leofeyer this is up4discussion again.

aschempp commented:
https://www.heise.de/mac-and-i/meldung/Apple-erklaert-Tracking-Cookies-den-Krieg-4330798.html

Interesting read in general. But as as summery, Google and other already use "First-party-cookies" (i.e. from your domain) to work around browser cookie limitations. Which means they all get sent to Contao, and would disable the reverse proxy.

m-vo commented:
Valid point! Maybe there should be some sort of whitelisting? - a tracking cookie would not interfere with what content to deliver (although it could theoretically be used for that), so why not let the devs decide?

Toflar committed:
Revamped concept

Toflar committed:
I guess we should be case insensitive

Toflar committed:
If the request is not cacheable we don't have to do any regex magic because it's irrelevant anyway

Toflar committed:
Added Matomo to the cookie blacklist

Toflar committed:
Added tests for the CacheListener

Toflar committed:
Added tests to ensure the listener is correctly registered

Toflar committed:
Added missing test for ContaoCache

Toflar committed:
Fixed tests on lowest deps

Toflar committed:
Test COOKIE_WHITELIST env variable

Toflar commented:
@contao/developers implemented as discussed. Coverage at 100% already 😄

leofeyer commented on core-bundle/src/EventListener/HttpCache/StripCookiesSubscriber.php:
```diff
@@ -0,0 +1,106 @@
+        '^_pk_ses$',
+        '^_pk_cvar$',
+        '^_pk_hsr$',
+    ];
```
Here are some more cookies that are shipped with Contao:

 * `TS_TABLE_` (used in the `tablesort.js` file)
 * `BE_PAGE_OFFSET`
 * `TL_VIEW`

I have already seen your ticket https://github.com/contao/contao/issues/459 hereto, however, we cannot be sure that none of the extensions use those cookies, therefore we should blacklist them.

Also, as briefly discussed on Slack, we should blacklist the `__cfduid` cookie used by Cloudflare.

Toflar replied:
* `TS_TABLE_` - we cannot ignore that cookie because it would break the functionality. The cookie would never make it to the application if we ignore it. That's another candidate for localstorage then.
* `BE_PAGE_OFFSET` - same here, we cannot ignore as it would break back end scroll to functionality. As mentioned in https://github.com/contao/contao/issues/459 it needs to be reworked to use localstorage or at least restricted to `/contao` so it's not in the FE. But we cannot ignore it as it would not get to the application then.
* `TL_VIEW` - same here again, if we would ignore it, the mobile layout functionality would be broken. You don't have any cache anyway if you use the `mobile-layout-bundle` so there's no point in ignoring it.

I'll add `__cfduid` though.

ausi commented on core-bundle/src/EventListener/HttpCache/StripCookiesSubscriber.php:
```diff
@@ -0,0 +1,106 @@
+
+        foreach ($cookies as $name) {
+            foreach ($list as $entry) {
+                $matches = preg_match('/'.$entry.'/i', $name);
```
I would add `^` and `$` here and remove them from the white/blacklist. This should make it less error-prone because nobody can forget to add them...

```suggestion
                $matches = preg_match('/^'.$entry.'$/i', $name);
```

Toflar replied:
Done in https://github.com/contao/contao/pull/389/commits/64c37a3ce0e667ea196ea8b980c21a568c8b44c2

Toflar committed:
Make sure start and end of a line are always matched

Toflar committed:
Added cloudflare cookie to the blacklist

Toflar committed:
Cleanup

Toflar committed:
Enable cache tracing for the upcoming Symfony 4.3 changes

Toflar commented on core-bundle/src/EventListener/HttpCache/StripCookiesSubscriber.php:
```diff
@@ -0,0 +1,109 @@
+
+        foreach ($cookies as $name) {
+            foreach ($list as $entry) {
+                $matches = preg_match('/^'.$entry.'$/i', $name);
```
@ausi maybe we could create one big regex of all regexes to see if any of them matches rather than checking them one after the other. Do you think that would be faster? 🤔

ausi replied:
Yes, should be faster I think. Can be done with a single `preg_grep()`:

```php
$removeCookies = preg_grep(
	'/^(?:'.implode(')$|^(?:', $list).')$/i',
	array_keys($request->cookies->all()),
	$isWhitelist ? PREG_GREP_INVERT : 0
);
foreach ($removeCookies as $name) {
	$request->cookies->remove($name);
}
```

leofeyer replied:
🦊

Toflar replied:
Jup, that was exactly what I was looking for. Just wasn't sure if this is going to be faster but I'm pretty confident it is. So implemented in da5b3904ee97d5bb6a8a8bd60a5c66f707f2e0e4.

Toflar committed:
Added facebook pixel cookie to the blacklist

Toflar committed:
Generate one big regular expression rather than checking each rule individually

aschempp commented on core-bundle/src/EventListener/CacheListener.php:
```diff
@@ -0,0 +1,60 @@
+        $response = $event->getResponse();
+
+        // 1. If the session was started.
+        if (null !== ($session = $request->getSession()) && $session->isStarted()) {
```
Doesn't this duplicate the Symfony session listener (https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/EventListener/AbstractSessionListener.php#L85) ? And if not, should we not use the same config (max-age 0 etc.)? Or even more so completely replace the `Cache-Control` header?

Toflar replied:
We know that if the session was started, we want to ensure that for Contao the response is private. We don't want to replace any caching headers because maybe things can be cached on client side, who knows? We just want to ensure that the reverse proxy is bypassed and I think it's good we've duplicated this logic because we're independent from what Symfony does in the next release.

aschempp commented on core-bundle/src/EventListener/CacheListener.php:
```diff
@@ -0,0 +1,60 @@
+
+        // 2. If the response sets a cookie (same reason as 1 but for other cookies than the session cookie).
+        if (0 !== \count($response->headers->getCookies())) {
+            $response->setPrivate();
```
As mentioned above, shouldn't it be `$response->headers->set('Cache-Control', 'private')`? Otherwise any max-age or similar headers would be kept. However, we would also remove any `no-cache` headers…

Toflar replied:
Same answer here.

aschempp commented on core-bundle/src/EventListener/HttpCache/StripCookiesSubscriber.php:
```diff
@@ -0,0 +1,106 @@
+class StripCookiesSubscriber implements EventSubscriberInterface
+{
+    private const BLACKLIST = [
+        // Modals are always for JS only
```
where does this come from? Can we add an example from the core?

Toflar replied:
There is no example, I've added it myself because I've checked my setups and I found one that sets a cookie named `myModal1` (some theme shit).

aschempp commented on core-bundle/src/Resources/contao/models/PageModel.php:
```diff
@@ -859,6 +860,7 @@ public function loadDetails()
 		$this->groups = $this->protected ? StringUtil::deserialize($this->groups) : false;
 		$this->layout = $this->includeLayout ? $this->layout : false;
 		$this->cache = $this->includeCache ? $this->cache : false;
+		$this->alwaysLoadFromCache = (bool) $this->alwaysLoadFromCache;
```
You must still check for `$this->includeCache` here, otherwise the respective checkbox could have been removed but the database entry would still affect the page settings.

Toflar replied:
True, fixed in 16ec76d379df8302857b2e0e40f8d405a2f7c39b.

aschempp commented on core-bundle/src/Resources/contao/models/PageModel.php:
```diff
@@ -912,6 +914,7 @@ public function loadDetails()
 					if ($objParentPage->includeCache)
 					{
 						$this->cache = $this->cache !== false ? $this->cache : $objParentPage->cache;
+						$this->alwaysLoadFromCache = $this->alwaysLoadFromCache !== false ? $this->alwaysLoadFromCache : $objParentPage->alwaysLoadFromCache;
```
Can probably be simplified to `$this->alwaysLoadFromCache = $this->alwaysLoadFromCache || $objParentPage->alwaysLoadFromCache;`

Toflar replied:
I wanted to stick to the coding style of all the other properties.

aschempp commented on manager-bundle/src/HttpKernel/ContaoCache.php:
```diff
@@ -82,21 +77,4 @@ protected function createStore()
-     */
-    private function isRequestPrivate(Request $request): bool
-    {
-        if ($request->headers->has('Authorization')) {
```
Are we no longer handling `Authorization` headers? Maybe we should check for them as well?

Toflar replied:
Yeah I wasn't sure. If so, it should go to the `CacheListener` (or whatever it's going to be named). What do you think @ausi @leofeyer ?

aschempp replied:
ping @leofeyer @ausi

leofeyer replied:
An `Authorization` header sure makes the request uncacheable, doesn't it? So we should handle it IMHO.

Toflar replied:
No we should not handle it. There is no case in Contao where you can restrict content to an `Authorization` header being present. The only case I can imagine is some `.htaccess` protection during development and that's exactly where it is irrelevant for us. So if someone really wants to build a module/fragment that needs this header, they must also `Vary` on it.

aschempp commented on core-bundle/src/EventListener/HttpCache/StripCookiesSubscriber.php:
```diff
@@ -0,0 +1,106 @@
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+class StripCookiesSubscriber implements EventSubscriberInterface
```
Every other class is called a `…Listener`, even if the subscribe…

Toflar replied:
No, all FOS cache events are subscribers. They were subscribers before too (header-replay).

aschempp commented on core-bundle/src/EventListener/CacheListener.php:
```diff
@@ -0,0 +1,60 @@
+
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+
+class CacheListener
```
This class name does not really explain what the listener does…

Toflar replied:
Yeah, I figured `CacheListener` is better than `TurnResponseIntoPrivateResponseIfOneOfTheThreeDocumentedReasonsIsGivenListener`. If you review then please suggest alternatives.

Toflar replied:
Proposals?

leofeyer replied:
`MakeResponsePrivateListener`?

Toflar replied:
Renamed in https://github.com/contao/contao/pull/389/commits/2adefcaef9cfa6cccb52edaf2d3e5f2714ddf022.

aschempp replied:
I find an "action naming" strage, I would have chosen `PrivateResponseListener` if that's what it's about. Or just `PrivateCacheListener`. However, is the response actually "private" or is it "no-cache" if there e.g. is a session? So maybe rather `UncacheableResponseListener`?

Toflar replied:
The response becomes private, which is correct. It may still be cached by the client, we don't care. So `UncacheableResponseListener` would be wrong. And it makes the response private but the listener itself is not so `PrivateCacheListener` sounds wrong too. It sounds like it's a listener that is only there for the private cache but it's exactly the opposite 😄 It's there to ensure that certain responses are never cached by any reverse proxy even if the developer said so.

Toflar committed:
Consider includeCache flag when considering alwaysLoadFromCache flag

Toflar committed:
Added a hidden image to allow to update cookies even if the page is served from the cache

Toflar committed:
Renamed the listener

aschempp commented on core-bundle/src/Controller/FrontendController.php:
```diff
@@ -101,4 +101,26 @@ public function logoutAction(): void
+     *
+     * @Route("/_contao/check_cookies", name="contao_frontend_check_cookies")
+     */
+    public function checkCookiesAction(): Response
```
...Action naming is deprecated afaik. Imho we should use an `__invoke`able controller for each action.

Toflar replied:
Not in this case, I just added a new action to an already existing controller. I don't think it's worth haviing multiple classes here. And even if we want to do so, then not in this PR.

aschempp commented on core-bundle/src/Controller/FrontendController.php:
```diff
@@ -101,4 +101,26 @@ public function logoutAction(): void
+        $response->setPrivate();
+        $response->headers->set('Content-Type', 'image/png');
+        $response->headers->addCacheControlDirective('no-store');
+        $response->headers->addCacheControlDirective('must-revalidate');
```
re-validation without expiration information does not make sense to me. I would suggest
```
$response->headers->set('Cache-Control', 'no-cache, no-store');
```

Toflar replied:
`no-store` is already present? Then we don't need `no-cache`, do we?

aschempp replied:
dunno, why would we need `must-revalidate` then? 😂

Toflar replied:
https://stackoverflow.com/questions/18148884/difference-between-no-cache-and-must-revalidate

> `max-age=0, must-revalidate` and `no-cache` aren't exactly identical. With `must-revalidate`, if the server doesn't respond to a revalidation request, the browser/proxy is supposed to return a 504 error. With `no-cache`, it would just show the cached content, which would be probably preferred by the user (better to have something stale than nothing at all). This is why `must-revalidate` is intended for critical transactions only.

I don't think that serving a cached response in case of an error makes sense.

aschempp commented on core-bundle/src/EventListener/HttpCache/StripCookiesSubscriber.php:
```diff
@@ -0,0 +1,106 @@
+        'AMP_TOKEN',
+        '_gac_.+',
+
+        // Matomo (https://matomo.org/faq/general/faq_146/)
```
Could we just use `_pk_.+` here?

Toflar replied:
We could but as Matomo specifies exactly which ones they set, I did take this list. I think it's safer. Otherwise we could also argue to just ignore `_.+` which would reduce the list quite a bit 😄

aschempp commented on core-bundle/src/EventListener/HttpCache/StripCookiesSubscriber.php:
```diff
@@ -0,0 +1,106 @@
+            return;
+        }
+
+        if ($this->whitelist) {
```
Should we use `!empty()` here?

Toflar replied:
Nah, very bad practice imho. We just do that in old code. `empty()` should never be used as it's not type safe. I'd use `0 !== count($this->whitelist)` here if this needs to be changed.

aschempp replied:
`if ($this->whitelist)` is not type-safe either?

Toflar replied:
It now is :)

aschempp commented on core-bundle/src/EventListener/MakeResponsePrivateListener.php:
```diff
@@ -0,0 +1,60 @@
+class MakeResponsePrivateListener
+{
+    /**
+     * Make sure the current response becomes a private response if any of the following
```
What if the request was not cacheable? Then it should not become private, right? Can we use `$response->isCacheable()` ?

Toflar replied:
If it's not cacheable we can ignore it anyway. Added in 66ee53af766e67ccb270488539d1399e5893541c.

aschempp commented on core-bundle/src/Resources/contao/classes/FrontendTemplate.php:
```diff
@@ -348,40 +348,53 @@ private function setCacheHeaders(Response $response)
 		/** @var PageModel $objPage */
 		global $objPage;
 
+		// Do not cache response if caching was not configured at all and make sure the response is private.
```
> was not configured at all

Thats not correct, the value is `0` if it **was** configured to be disabled.

Toflar replied:
It's still correct. If it was not configured it's false. Extended the comment in ea3091400445e9455b6813e3fe7d83e221bacb81.

aschempp commented on core-bundle/src/Resources/contao/dca/tl_page.php:
```diff
@@ -181,7 +181,7 @@
 		'createSitemap'               => 'sitemapName',
 		'includeLayout'               => 'layout',
-		'includeCache'                => 'cache,clientCache',
+		'includeCache'                => 'cache,alwaysLoadFromCache,clientCache',
```
Adding the checkbox after both dropdowns gives a better look&feel imho

Toflar replied:
I don't think that's true because the checkbox belongs to the shared cache only. This is how it looks now:

<img width="1190" alt="Bildschirmfoto 2019-06-03 um 13 05 00" src="https://user-images.githubusercontent.com/481937/58797402-36dbb080-8600-11e9-8d16-4038a65643c3.png">

leofeyer replied:
Haven't we agreed on not having a checkbox at all?

Toflar replied:
No, I said I'm going to try and if I fail, we'll remove it. But https://github.com/contao/contao/pull/483 must be merged to make it work so you should provide feedback there :)

leofeyer replied:
AFAIR we have agreed not to add the checkbox, because we cannot foresee which problems it will cause (although we can foresee that it will make everything a lot more complex). Do we need to rediscuss?

Toflar replied:
It's funny how people always keep saying "we agreed". Maybe things we discuss on Mumble need to be documented better because the way I remember it is that we really do want to have this checkbox and if we're able to find a solution that is fairly simple to understand and implement, we'll do it. Only if there's no simple way to get it working, we'll remove it. I then said I'll have another look at it which I did and this is the result.

I cannot imagine any side effects other than people **not** getting personal content if they should. This would be a misconfiguration then and is not relevant for security. The other way around (people getting personal content of others) would be security relevant but this not possible and remains unaffected by this checkbox.

So I would definitely have it merged the way it is so @ausi can go ahead and test his use case and see if it even works with the remember me (which I'm confident it will together with #483 ) 😄

leofeyer replied:
@aschempp @ausi What do you recall?

aschempp replied:
We discussed to not add it because there was no easy solution. But then we found one, so this implementation certainly is right 🙃

leofeyer replied:
That's great. Then you should probably share your solution with us, so we can discuss it altogether.

aschempp replied:
The solution is already implemented 😂 it's the one-pixel image request, which will fetch the cookies.

ausi replied:
> the way I remember it is that we really do want to have this checkbox and if we're able to find a solution that is fairly simple to understand and implement, we'll do it.

Yes, AFAIR this was the latest state of our discussion.

aschempp commented on core-bundle/src/Resources/contao/dca/tl_page.php:
```diff
@@ -524,6 +524,13 @@
+			'label'                   => &$GLOBALS['TL_LANG']['tl_page']['alwaysLoadFromCache'],
+			'exclude'                 => true,
+			'inputType'               => 'checkbox',
+			'sql'                     => "char(1) NOT NULL default ''"
```
add `eval => tl_class => 'w50 m12'`

Toflar replied:
Added in d8cb92b62d099b0c7ad7d65f00835d18dc6616b2.

aschempp commented on core-bundle/src/Resources/contao/languages/en/tl_page.xlf:
```diff
@@ -171,13 +171,19 @@
+        <source>Always load from shared cache (server cache)</source>
+      </trans-unit>
+      <trans-unit id="tl_page.alwaysLoadFromCache.1">
+        <source>Always load this page from the shared cache, even if a member is logged in. (Note that you can no longer personalize the page for logged in members in this case.)</source>
```
No brackets needed imho

Toflar replied:
True, updated in 871b9b13373b40b7bba0f70efc4f2ded10b52895.

aschempp commented on core-bundle/src/Resources/contao/models/PageModel.php:
```diff
@@ -859,6 +860,7 @@ public function loadDetails()
 		$this->groups = $this->protected ? StringUtil::deserialize($this->groups) : false;
 		$this->layout = $this->includeLayout ? $this->layout : false;
 		$this->cache = $this->includeCache ? $this->cache : false;
+		$this->alwaysLoadFromCache = $this->includeCache ? $this->alwaysLoadFromCache : $this->alwaysLoadFromCache;
```
Shouldn't the else be `false`?

Toflar replied:
Lol, of course! Good catch, fixed in 22e224d0a71e4aca278c6d76df3648f600746869.

aschempp commented on core-bundle/src/Resources/contao/pages/PageRegular.php:
```diff
@@ -481,6 +481,13 @@ protected function createTemplate($objPage, $objLayout)
+		// checkbox is configured
+		$this->Template->checkCookiesUrl = '';
+		if ($objPage->alwaysLoadFromCache) {
+			$this->Template->checkCookiesUrl = System::getContainer()->get('router')->generate('contao_frontend_check_cookies');
```
Instead of creating an actual route that should not be used because its *internal*, could we use a controller reference throught `_fragments` and sign the request?

Toflar replied:
I don't think we should do that. This is a public route so it doesn't need signing. If you want to call that route, feel free to do so, who cares. It's not internal at all. Or did you mean "internal" because nobody else should be using it? Well then, so be it. If people use it then they do...you're not supposed to use anything marked with `@internal` either but you still can. I think it's correct the way it is.

aschempp replied:
it just also "pollutes" the routing 🤷‍♂

aschempp commented on manager-bundle/src/HttpKernel/ContaoCache.php:
```diff
@@ -31,45 +31,40 @@ class ContaoCache extends HttpCache implements CacheInvalidation
         $this->addSubscriber(new CleanupCacheTagsListener());
     }
 
-    public function wasBypassed(): bool
```
This would be an API break, or was this not yet released?

Toflar replied:
Was not released yet. It's only on master.

Toflar committed:
Do not make the response private if it is not cacheable for a reverse proxy anyway

Toflar committed:
Improved docs

Toflar committed:
Added missing CSS classes

Toflar committed:
Updated description

Toflar committed:
Fixed variable assignment

Toflar committed:
CS

leofeyer commented on manager-bundle/src/HttpKernel/ContaoCache.php:
```diff
@@ -31,45 +31,40 @@ class ContaoCache extends HttpCache implements CacheInvalidation
 
+        $this->isDebug = $kernel->isDebug();
+
+        $this->addSubscriber(new StripCookiesSubscriber(array_filter(explode(',', getenv('COOKIE_WHITELIST') ?: ''))));
```
Is this meant to be `?:` or `??`?

Toflar replied:
No `?:` is correct. `getenv()` returns `false` if it isn't set. `??` is short for `isset()` which makes no sense here.

leofeyer committed:
Fix the coding style

leofeyer commented on core-bundle/src/Resources/contao/classes/FrontendTemplate.php:
```diff
@@ -348,40 +348,62 @@ private function setCacheHeaders(Response $response)
+			//
+			// Exception to this rule is if those modules support ESI requests,
+			// in which case they can handle their own caching headers and only
+			// set "Cache-Control: public" if no member is logged in.
```
This explanation is kind of longish and not necessarily on the spot (no hard feelings). Maybe we can rework it?

Toflar replied:
Proposals? It is complex so it should be long and detailed :)

leofeyer replied:
```
CURRENT:
    // We do vary on cookie for the responses that are cacheable for
    // the shared cache, which will cause reverse proxies to never load
    // a response from the cache if the _request_ contains a cookie on
    // the next request.

PROPOSED:
    // We vary on cookies if a response is cacheable by
    // the shared cache, so a reverse proxy does not load
    // a response from the cache if the _request_ contains a cookie.


CURRENT:
    // NOTE: We do not generate a cache entry for every response
    // containing a cookie! Responses containing a cookie will always
    // be private (see earlier in this method).

PROPOSED:
    // This DOES NOT mean that we generate a cache entry for every response
    // containing a cookie! Responses containing cookies will always
    // be private (see earlier in this method).


CURRENT:
    // However, we want to be able to force the reverse proxy to still
    // load the response from the cache even if there is a cookie on
    // the request in case it was configured so by the user. A typical
    // example for this would be if you have a member section in a
    // completely different page tree than your "public" website.
    //
    // You can configure the public website to force the cache, which
    // will ensure that logged in members still get the cached entries
    // served when they visit it. However, you have to be very careful
    // in this case, e.g. you cannot share any front end module that
    // shows different content depending on whether you are logged in
    // or not in both trees (such as a navigation or login module).
    //
    // Exception to this rule is if those modules support ESI requests,
    // in which case they can handle their own caching headers and only
    // set "Cache-Control: public" if no member is logged in.

PROPOSED:
    // However, we want to be able to force the reverse proxy to still
    // load a response from cache – even if there is a cookie on
    // the request – in case the admin has configured it so. A typical
    // example for this would be a protected member section in a
    // completely different page tree than the public website.
    // You can configure the public website to force loading from cache, which
    // ensures that logged in members still get served the cached entries.
    //

(The next paragraph is the most confusing one IMHO.)
    // However, you have to be very careful
    // in this case! You e.g. cannot share a front end module that
    // shows different content depending on whether you are logged in
    // or not in both trees (such as a navigation or login module).
    //
    // An exception to this rule is if those modules support ESI requests,
    // in which case they can handle their own caching headers and only
    // set "Cache-Control: public" if no member is logged in.
```

Toflar replied:
I've updated and shortened the comment. Wdyt?

leofeyer commented on core-bundle/src/Resources/contao/templates/frontend/fe_page.html5:
```diff
@@ -94,5 +94,8 @@
   <?= $this->mootools ?>
 
+  <?php if ($this->checkCookiesUrl): ?>
+    <img src="<?= $this->checkCookiesUrl ?>" width="1" height="1" class="invisible" alt role="presentation">
```
Why did you add the `role="presentation"` attribute here?

Toflar replied:
For screen readers. I read up on the topic and apparently most of them ignore images that have an `alt` attribute set to `null` but I figured we can just be very sure and add `role` too :)

leofeyer replied:
So why not `aria-hidden`?

Toflar replied:
Hmm according to https://timwright.org/blog/2016/11/19/difference-rolepresentation-aria-hiddentrue/ it would actually be even better. Changed in 8ffc3194d8949a1240bd5aeb3cc98f1928688abb

leofeyer commented:
Here is what needs to be done so I can merge the PR:

* [x] Discuss the wording (see below)
* [x] Rework the longish explanation
* [x] Fix the three failing unit tests
* [x] Address my latest comments

Regarding the wording: I don't like that we are now using "shared cache (server cache)" and "private cache (client cache)". Is this really necessary? If "shared cache" and "private cache" are the "correct" terms, we should use only them IMHO.

Toflar committed:
Fixed tests

Toflar commented:
Tests fixed, comments addressed.

Toflar commented:
> Regarding the wording: I don't like that we are now using "shared cache (server cache)" and "private cache (client cache)". Is this really necessary? If "shared cache" and "private cache" are the "correct" terms, we should use only them IMHO.

I just figured a lot of people are probably not familiar with private and public cache but they can sort of imagine client and server cache. But I'm also fine with only using the correct terms.

Toflar committed:
Use aria-hidden="true" instead of role="presentation"

Toflar committed:
Reworked caching explanation

Toflar committed:
Added a test case for the new check cookies action

leofeyer committed:
Rework the long comment again

leofeyer committed:
Remove the old "server cache" and "client cache" terms

leofeyer commented:
Thank you very much @Toflar!
